crypto/cms/cms_smime.c

```diff
@@ -631,6 +631,7 @@ int CMS_decrypt_set1_pkey(CMS_ContentInfo *cms, EVP_PKEY *pk, X509 *cert)
          * all.
          */
         else if (!cert || !CMS_RecipientInfo_ktri_cert_cmp(ri, cert)) {
+            EVP_PKEY_up_ref(pk);
```

Comment:
This kind of calls for question why isn't it appropriate to upref in CMS_RecipientInfo_set0_pkey. I mean formally speaking upref belongs next to pointer assignment, it's an expression of "I'm holding reference too".

Reply:
Because it's a "set0" function. If it were a "set1" function I would agree with you. However "set0" implies that ownership of the reference is passed along with the function call. In a sense, for a set0, you can view the "assignment" as being the call site for the function - which is where this up-ref is.

Reply:
So you mean that it's reflection of the fact that "set0" is called from "set1"... It's still quite counter-intuitive pattern, to demand that object has to be properly upref-ed by *caller*. Well, it's also counter-intuitive to transfer ownership to private key. I mean it's the kind of thing application should definitely hold on... I mean on certain level one can wonder if there should be any "set0" with private key as argument...

Reply:
The set0/set1 pattern has absolutely nothing to do with what type of object it is - and there is nothing counter-intuitive to me to have a private key "ownership" pass into a crypto API ... 

Who is responsible for freeing an item needs to be clear explicitly - and that is precisely what this API pattern is there to indicate (compared to the prior pattern of whatever-the-function-does-and-you-just-have-to-know-which-it-is approach).